cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		return err
+	}
+	// TODO: revisit how sample nodes are chosen.
+	for _, node := range nodes {
+		if belongs, err := mig.Belongs(node); err == nil && belongs {
+			mig.sampleNode = node
+			break
+		}
+	}
+	if mig.sampleNode == nil {
+		return fmt.Errorf("no sample node found for %s", mig.Id())
+	}
+	gce.migs = append(gce.migs, mig)
+	gce.gceManager.RegisterMig(mig)
+	return nil
+}
+
+// Name returns name of the cloud provider.
+func (gce *GceCloudProvider) Name() string {
+	return "gce"
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (gce *GceCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	result := make([]cloudprovider.NodeGroup, 0, len(gce.migs))
+	for _, mig := range gce.migs {
+		result = append(result, mig)
+	}
+	return result
+}
+
+// NodeGroupForNode returns the node group for the given node.
+func (gce *GceCloudProvider) NodeGroupForNode(node *kube_api.Node) (cloudprovider.NodeGroup, error) {
+	ref, err := GceRefFromProviderId(node.Spec.ProviderID)
+	if err != nil {
+		return nil, err
+	}
+	mig, err := gce.gceManager.GetMigForInstance(ref)
+	return mig, err
+}
+
+// GceRef contains s reference to some entity in GCE/GKE world.
+type GceRef struct {
+	Project string
+	Zone    string
+	Name    string
+}
+
+// GceRefFromProviderId creates InstanceConfig object
+// from provider id which must be in format:
+// gce://<project-id>/<zone>/<name>
+// TODO(piosz): add better check whether the id is correct
+func GceRefFromProviderId(id string) (*GceRef, error) {
+	splitted := strings.Split(id[6:], "/")
+	if len(splitted) != 3 {
+		return nil, fmt.Errorf("Wrong id: expected format gce://<project-id>/<zone>/<name>, got %v", id)
+	}
+	return &GceRef{
+		Project: splitted[0],
+		Zone:    splitted[1],
+		Name:    splitted[2],
+	}, nil
+}
+
+// Mig implements NodeGroup interfrace.
+type Mig struct {
+	GceRef
+
+	gceManager *GceManager
+
+	minSize    int
+	maxSize    int
+	sampleNode *kube_api.Node
+}
+
+// MaxSize returns maximum size of the node group.
+func (mig *Mig) MaxSize() int {
+	return mig.maxSize
+}
+
+// MinSize returns minimum size of the node group.
+func (mig *Mig) MinSize() int {
+	return mig.minSize
+}
+
+// Size returns the current TARGET size of the node group. It is possible that the
+// number is different from the number of nodes registered in Kuberentes.
+func (mig *Mig) Size() (int, error) {
+	size, err := mig.gceManager.GetMigSize(mig)
+	return int(size), err
+}
+
+// SampleNode returns a sample node for the mig. Assumes that mig definition doesn't change over time.
+// The node may not exist anymore.
+func (mig *Mig) SampleNode() (*kube_api.Node, error) {
+	if mig.sampleNode != nil {
+		return mig.sampleNode, nil
+	}
+	return nil, fmt.Errorf("no sample node available")
+}
+
+// IncreaseSize increases Mig size
+func (mig *Mig) IncreaseSize(delta int) error {
+	if delta <= 0 {
+		return fmt.Errorf("size increase must be positive")
```

Comment:
suppose that some has -1 in an int variable. Stupid int->uint conversion would make it some huge number that added to the current size couple lines below would effectively mean -1. In this way -1 would pass through the checks.